### ISSUE TITLE
Don't use PR for version bumps?

### ISSUE THREAD

**Issue by Kagami** [PistonDevelopers/piston#1003]
*Continue discussion from https://github.com/PistonDevelopers/gfx_text/pull/33*

>In my opinion creating PR just for version bump is wrong.
> * git tag should point to exact commit, not to the merge
> * cargo publish should be made from exact same commit as git tag
> * version should be bumped only after everything was tested (e.g. docs, README formatting, etc), so it can't be part of another PR; and creating PR only for single line change is pointless

As far as I understand, the main reason for using PR for version bumps is what you'll get email. Is it possible to achieve that in another way? Maybe via some push hook?

cc @bvssvni

**Comment by bvssvni** [comment]
I am not sure what you mean with git tags and why cargo publish should be made from the exact same commit. Can you explain the idea to me?

**Comment by Kagami** [comment]
It's a sort of convention in (I believe) a lot of languages. You have some file with metadata of your project (setup.py, package.json, project_name.cabal, etc) and when you make a new release, you update some number in that file. Since generally only that number separate one project version from another, it's a good idea to make it as separate commit to main branch and tag it right away. And if you publish your package to the language repo (crates.io, pypi, hackage, etc) from that commit, all 3 things (commit, tag, release) will be perfectly linked together.

**Comment by bvssvni** [comment]
Here is what I do:

1. Bump version locally.
2. Run tests and examples to see if it works.
3. Make a PR.
4. Merge.
5. Publish from local repo.

So, the library is published immediately after the version is merged. People who want to know the exact commit can check the history.

I don't make git tags because this requires extra work.

The problem is that even with Eco to do the analysis, it takes hours to update the entire ecosystem. Any extra steps slows down the progress. What I care about is to get the ecosystem integrated and do it fast enough to be able to work on other stuff.

**Comment by bvssvni** [comment]
It is worth considering that the Rust ecosystem is pretty unstable at the moment, and that we might change the practice when libraries get more mature.

**Comment by Kagami** [comment]
Ok, if that works for you, I guess it's not hard for me to create extra PR.

I just don't fully understand, if eco is an extra tool, can't it periodically fetch latest package versions in some way?

**Comment by bvssvni** [comment]
Eco is almost smart enough, but there are some cases that we need intervention and control:

- Some breaking changes require fixes to the code
- Some libraries are external and not under the control of PistonDevelopers
- Whether other people are working on a PR and we might want to merge it first
- Testing for runtime errors requires executing examples to see if it works
- Dev-dependencies are tricky, might require an additional layer of analysis
- The update logic can be unsound with "holes" in the ecosystem. This can occur when adding a new library. Eco could compare dependency graph with crates.io, but need to handle all edge cases.

The plan is to improve Eco bit by bit over time and possibly automate more. However, it is easier to check what Eco is recommending instead of trusting the ecosystem information to be correct. This can be done while updating the ecosystem and handle all the cases above.

**Comment by Kagami** [comment]
Ok. Though I think it's better to indicate version bumps more explicitly than via PR? I mean, if you depend on something, you're not interested in common non-breaking changes, you just want to know when library bumps its major number.

**Comment by bvssvni** [comment]
Change in the first non-zero number means a breaking change and should only be increased when:

- The library has a breaking change
- Any of the dependencies bumps the first non-zero number

The reason we need to bump when any of the dependency changes: Cargo attempts to link two different versions of the same library which causes a compiler error. If you don't bump the first non-zero number, but a second non-zero number, `cargo update` will break existing code. Eco tells the correct version to bump to in order to keep the ecosystem integrated.

**Comment by Kagami** [comment]
I mean if people want to get email about version bumps, they need to watch the repo. But then you'll get emails about all PRs, not just version changes. So I'm wondering, is there some better way to notify people about breaking changes in the library.

**Comment by bvssvni** [comment]
Yes, this could be better, but usually there is one person doing the updates. We want to avoid waiting for people to read their emails and doing updates themselves before we get the ecosystem integrated. I am not sure what benefit it would be to receive an email per library. If you have 20 libraries, you get 20 emails.

The simplest way to get this information is by running Eco. You just create a list like [this one](https://github.com/PistonDevelopers/eco/blob/master/assets/extract/piston.txt) and write a [small program](https://github.com/PistonDevelopers/eco/blob/master/examples/piston.rs) to print out the updates.
